clippy_lints/src/paths_from_format.rs

```diff
@@ -0,0 +1,132 @@
+                    .iter()
+                    .map(|x| Sugg::hir_with_applicability(cx, x.param.value, "..", &mut applicability))
+                    .collect();
+                let mut paths_zip = string_parts.iter().take(real_vars.len()).zip(real_vars.clone());
+                let mut sugg = String::new();
+                if let Some((part, arg)) = paths_zip.next() {
+                    if is_valid_use_case(string_parts.first().unwrap_or(&""), string_parts.get(1).unwrap_or(&"")) {
+                        return;
+                    }
+                    if part.is_empty() {
+                        sugg = format!("Path::new(&{arg})");
+                    }
+                    else {
+                        push_comps(&mut sugg, part);
+                        let _ = write!(sugg, ".join(&{arg})");
+                    }
+                }
+                for n in 1..real_vars.len() {
+                    if let Some((part, arg)) = paths_zip.next() {
+                        if is_valid_use_case(string_parts.get(n).unwrap_or(&""), string_parts.get(n+1).unwrap_or(&"")) {
+                            return;
+                        }
+                        else if n < real_vars.len() {
+                            push_comps(&mut sugg, part);
+                            let _ = write!(sugg, ".join(&{arg})");
+                        }
+                        else {
+                            sugg = format!("{sugg}.join(&{arg})");
+                        }
+                    }
+                }
+                if real_vars.len() < string_parts.len() {
+                    push_comps(&mut sugg, string_parts[real_vars.len()]);
+                }
+                span_lint_and_sugg(
+                    cx,
+                    PATHS_FROM_FORMAT,
+                    expr.span,
+                    "`format!(..)` used to form `PathBuf`",
+                    "consider using `Path::new()` and `.join()` to make it OS-agnostic and improve code readability",
+                    sugg,
+                    Applicability::MaybeIncorrect,
+                );
+            }
+        }
+    }
+}
+
+fn push_comps(string: &mut String, path: &str) {
+    let mut path = path.to_string();
+    if !string.is_empty() {
+        path = path.trim_start_matches(|c| c == '\\' || c == '/').to_string();
+    }
+    for n in Path::new(&path).components() {
+        let mut x = n.as_os_str().to_string_lossy().to_string();
+        if string.is_empty() {
+            let _ = write!(string, "Path::new(\"{x}\")");
+        } else {
+            x = x.trim_end_matches(|c| c == '/' || c == '\\').to_string();
+            let _ = write!(string, ".join(\"{x}\")");
+        }
+    }
+}
+
+// In essence, this checks if the format!() used is made for concatenation of the filenames / folder
+// names itself. This returns true when it is something like
+// `PathBuf::from(format!("/x/folder{}/textfile.txt", folder_number))`
```

Comment:
Based on the comment in the related issue, I would actually think that such a paht can still be problematic. I'd suggest either, checking if the path starts with `//?/` which makes it a verbatim path under windows or (IMO the better approach) add a config value, to allow this. I have the feeling the lint can be very noisy without, but ignoring these cases can also be wrong.